### PR TITLE
Recover data written to wrong column families

### LINKED ISSUE #15804: `MultiTenancySignalSubscriptionStateMigration` fails with `IndexOutOfBoundsException`
The `MultiTenancySignalSubscriptionStateMigration` failed with the following exception:
```
 java.lang.IndexOutOfBoundsException: index=30 length=8 capacity=30
	at org.agrona.AbstractMutableDirectBuffer.boundsCheck0(AbstractMutableDirectBuffer.java:1719) ~[agrona-1.20.0.jar:1.20.0]
	at org.agrona.AbstractMutableDirectBuffer.getLong(AbstractMutableDirectBuffer.java:129) ~[agrona-1.20.0.jar:1.20.0]
	at io.camunda.zeebe.db.impl.DbLong.wrap(DbLong.java:27) ~[zeebe-db-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.db.impl.DbCompositeKey.wrap(DbCompositeKey.java:47) ~[zeebe-db-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.visit(TransactionalColumnFamily.java:450) ~[zeebe-db-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.lambda$forEachInPrefix$20(TransactionalColumnFamily.java:387) ~[zeebe-db-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.ColumnFamilyContext.withPrefixKey(ColumnFamilyContext.java:112) ~[zeebe-db-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.forEachInPrefix(TransactionalColumnFamily.java:371) ~[zeebe-db-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.forEachInPrefix(TransactionalColumnFamily.java:342) ~[zeebe-db-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.lambda$forEach$7(TransactionalColumnFamily.java:162) ~[zeebe-db-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.lambda$ensureInOpenTransaction$19(TransactionalColumnFamily.java:325) ~[zeebe-db-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.DefaultTransactionContext.runInTransaction(DefaultTransactionContext.java:31) ~[zeebe-db-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.ensureInOpenTransaction(TransactionalColumnFamily.java:324) ~[zeebe-db-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.forEach(TransactionalColumnFamily.java:160) ~[zeebe-db-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.engine.state.migration.to_8_4.DbSignalSubscriptionMigrationState.migrateSignalSubscriptionStateForMultiTenancy(DbSignalSubscriptionMigrationState.java:44) ~[zeebe-workflow-engine-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.engine.state.migration.DbMigrationState.migrateSignalSubscriptionStateForMultiTenancy(DbMigrationState.java:458) ~[zeebe-workflow-engine-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.engine.state.migration.to_8_4.MultiTenancySignalSubscriptionStateMigration.runMigration(MultiTenancySignalSubscriptionStateMigration.java:30) ~[zeebe-workflow-engine-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.engine.state.migration.DbMigratorImpl.runMigration(DbMigratorImpl.java:170) ~[zeebe-workflow-engine-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.engine.state.migration.DbMigratorImpl.handleMigrationTask(DbMigratorImpl.java:143) ~[zeebe-workflow-engine-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.engine.state.migration.DbMigratorImpl.lambda$runMigrations$1(DbMigratorImpl.java:93) ~[zeebe-workflow-engine-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.DefaultTransactionContext.runInNewTransaction(DefaultTransactionContext.java:61) ~[zeebe-db-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.DefaultTransactionContext.runInTransaction(DefaultTransactionContext.java:33) ~[zeebe-db-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.engine.state.migration.DbMigratorImpl.runMigrations(DbMigratorImpl.java:91) ~[zeebe-workflow-engine-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.broker.system.partitions.impl.steps.MigrationTransitionStep.transitionTo(MigrationTransitionStep.java:52) ~[zeebe-broker-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.broker.system.partitions.impl.PartitionTransitionProcess.lambda$proceedWithTransition$1(PartitionTransitionProcess.java:94) ~[zeebe-broker-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:94) [zeebe-scheduler-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:47) [zeebe-scheduler-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:120) [zeebe-scheduler-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:130) [zeebe-scheduler-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:108) [zeebe-scheduler-8.4.0.jar:8.4.0]
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:227) [zeebe-scheduler-8.4.0.jar:8.4.0] 
```

As a result the partition is unavailable. The exception occurred twice in a row on the same broker. The cluster has since been deleted so no data is available to reproduce.

Zeebe Version: 8.4.0

### PR DESCRIPTION
**Pull request by korthout**
## Description

Recovers data that was written to wrong column families, by moving it to the correct ones ([details](https://github.com/camunda/zeebe/issues/15804#issuecomment-1908128577)):
- 48: `DMN_DECISION_KEY_BY_DECISION_ID_AND_VERSION` to `MESSAGE_STATS`
- 49: `DMN_DECISION_REQUIREMENTS_KEY_BY_DECISION_REQUIREMENT_ID_AND_VERSION` to `PROCESS_INSTANCE_KEY_BY_DEFINITION_KEY`
- 50: `SIGNAL_SUBSCRIPTION_BY_NAME_AND_KEY` to `MIGRATIONS_STATE`

This data recovery happens as one of the migration tasks (after the to_8.3 and before the to_8.4 migration tasks). 

Data is detected as either fitting (by key/value length and ability to be deserialized) before being moved. This allows any correct entries to be left alone.

Both debug and trace logs have been added:
- Debug level logs for copying and deleting entries from column families
- Trace level logs for detection of data

Tests have been added to verify a few different cases, but perhaps not all edge cases are covered. If you see any, please let me know.

> [!NOTE]
> All column family references, both in tests as well as in the migration code, ignore foreign keys. The main reason is that some data is moved in parts. Column family 48 and 49 deal with DMN models in that one deals with decisions and the other with the decision requirements. These belong together when modeling the full picture, which means that foreign key constraints cannot be met until both correctors have been executed. I've also chosen to ignore foreign keys to avoid complicating the tests too much.

> [!NOTE]
> As a migration task, this means that the data is only recovered once a broker becomes the leader. Migration tasks are not executed on followers, as far as I know. I don't think this will lead to problems for the given data, but theoretically, it could. If you see a solution to circumvent this, I'd love to learn about it. For now, I think we have to take that risk.

> [!NOTE]
> While this fixes the bug on 8.4, it will be backported manually. That is because the impact of mistakes is still large when backporting such changes. We must first ensure that the Column Family cannot be accidentally changed again before we can backport this. I will follow up on this soon after this pull request is merged.

## Manual test

I've performed a manual test with the data folder from the related issue. These show how the migration tasks recover some of the faulty data and then continue past the migration task that led to the error in #15804.

<details><summary>Migration logs</summary>
<p>

```
18:26:52.264 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Starting processing of migration tasks (use LogLevel.DEBUG for more details) ... 
18:26:52.264 [Broker-0] [zb-actors-0] [ZeebePartition-1] DEBUG io.camunda.zeebe.engine.state.migration - Found 18 migration tasks: ProcessMessageSubscriptionSentTimeMigration, MessageSubscriptionSentTimeMigration, TemporaryVariableMigration, DecisionMigration, DecisionRequirementsMigration, ProcessInstanceByProcessDefinitionMigration, JobTimeoutCleanupMigration, JobBackoffCleanupMigration, MultiTenancyProcessStateMigration, MultiTenancyDecisionStateMigration, MultiTenancyMessageStateMigration, MultiTenancyMessageStartEventSubscriptionStateMigration, MultiTenancyMessageSubscriptionStateMigration, MultiTenancyProcessMessageSubscriptionStateMigration, MultiTenancyJobStateMigration, ColumnFamilyPrefixCorrectionMigration, MultiTenancySignalSubscriptionStateMigration, JobBackoffRestoreMigration
18:26:52.267 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Skipping ProcessMessageSubscriptionSentTimeMigration migration (1/18).  It was determined it does not need to run right now.
18:26:52.267 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Skipping MessageSubscriptionSentTimeMigration migration (2/18).  It was determined it does not need to run right now.
18:26:52.267 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Skipping TemporaryVariableMigration migration (3/18).  It was determined it does not need to run right now.
18:26:52.267 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Skipping DecisionMigration migration (4/18).  It was determined it does not need to run right now.
18:26:52.267 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Skipping DecisionRequirementsMigration migration (5/18).  It was determined it does not need to run right now.
18:26:52.267 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Starting ProcessInstanceByProcessDefinitionMigration migration (6/18)
18:26:52.268 [Broker-0] [zb-actors-0] [ZeebePartition-1] DEBUG io.camunda.zeebe.engine.state.migration - ProcessInstanceByProcessDefinitionMigration migration completed in 0 ms.
18:26:52.268 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Finished ProcessInstanceByProcessDefinitionMigration migration (6/18)
18:26:52.268 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Starting JobTimeoutCleanupMigration migration (7/18)
18:26:52.268 [Broker-0] [zb-actors-0] [ZeebePartition-1] DEBUG io.camunda.zeebe.engine.state.migration - JobTimeoutCleanupMigration migration completed in 0 ms.
18:26:52.268 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Finished JobTimeoutCleanupMigration migration (7/18)
18:26:52.268 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Starting JobBackoffCleanupMigration migration (8/18)
18:26:52.268 [Broker-0] [zb-actors-0] [ZeebePartition-1] DEBUG io.camunda.zeebe.engine.state.migration - JobBackoffCleanupMigration migration completed in 0 ms.
18:26:52.268 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Finished JobBackoffCleanupMigration migration (8/18)
18:26:52.268 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Skipping MultiTenancyProcessStateMigration migration (9/18).  It was determined it does not need to run right now.
18:26:52.268 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Skipping MultiTenancyDecisionStateMigration migration (10/18).  It was determined it does not need to run right now.
18:26:52.268 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Skipping MultiTenancyMessageStateMigration migration (11/18).  It was determined it does not need to run right now.
18:26:52.268 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Skipping MultiTenancyMessageStartEventSubscriptionStateMigration migration (12/18).  It was determined it does not need to run right now.
18:26:52.268 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Skipping MultiTenancyMessageSubscriptionStateMigration migration (13/18).  It was determined it does not need to run right now.
18:26:52.268 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Skipping MultiTenancyProcessMessageSubscriptionStateMigration migration (14/18).  It was determined it does not need to run right now.
18:26:52.269 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Skipping MultiTenancyJobStateMigration migration (15/18).  It was determined it does not need to run right now.
18:26:52.269 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Starting ColumnFamilyPrefixCorrectionMigration migration (16/18)
18:26:52.269 [Broker-0] [zb-actors-0] [ZeebePartition-1] DEBUG io.camunda.zeebe.engine.state.migration - Copying entry with key[DbByte{bytes=���JobBackoffCleanupMigration}] from column family [50] DEPRECATED_SIGNAL_SUBSCRIPTION_BY_NAME_AND_KEY to column family [56] MIGRATIONS_STATE
18:26:52.270 [Broker-0] [zb-actors-0] [ZeebePartition-1] DEBUG io.camunda.zeebe.engine.state.migration - Deleting entry with key[DbByte{bytes=���JobBackoffCleanupMigration}] from column family [50] DEPRECATED_SIGNAL_SUBSCRIPTION_BY_NAME_AND_KEY
18:26:52.270 [Broker-0] [zb-actors-0] [ZeebePartition-1] DEBUG io.camunda.zeebe.engine.state.migration - Copying entry with key[DbByte{bytes=���JobBackoffRestoreMigration}] from column family [50] DEPRECATED_SIGNAL_SUBSCRIPTION_BY_NAME_AND_KEY to column family [56] MIGRATIONS_STATE
18:26:52.270 [Broker-0] [zb-actors-0] [ZeebePartition-1] DEBUG io.camunda.zeebe.engine.state.migration - Deleting entry with key[DbByte{bytes=���JobBackoffRestoreMigration}] from column family [50] DEPRECATED_SIGNAL_SUBSCRIPTION_BY_NAME_AND_KEY
18:26:52.270 [Broker-0] [zb-actors-0] [ZeebePartition-1] DEBUG io.camunda.zeebe.engine.state.migration - Copying entry with key[DbByte{bytes=���JobTimeoutCleanupMigration}] from column family [50] DEPRECATED_SIGNAL_SUBSCRIPTION_BY_NAME_AND_KEY to column family [56] MIGRATIONS_STATE
18:26:52.270 [Broker-0] [zb-actors-0] [ZeebePartition-1] DEBUG io.camunda.zeebe.engine.state.migration - Deleting entry with key[DbByte{bytes=���JobTimeoutCleanupMigration}] from column family [50] DEPRECATED_SIGNAL_SUBSCRIPTION_BY_NAME_AND_KEY
18:26:52.270 [Broker-0] [zb-actors-0] [ZeebePartition-1] DEBUG io.camunda.zeebe.engine.state.migration - Copying entry with key[DbByte{bytes=���TemporaryVariableMigration}] from column family [50] DEPRECATED_SIGNAL_SUBSCRIPTION_BY_NAME_AND_KEY to column family [56] MIGRATIONS_STATE
18:26:52.270 [Broker-0] [zb-actors-0] [ZeebePartition-1] DEBUG io.camunda.zeebe.engine.state.migration - Deleting entry with key[DbByte{bytes=���TemporaryVariableMigration}] from column family [50] DEPRECATED_SIGNAL_SUBSCRIPTION_BY_NAME_AND_KEY
18:26:52.270 [Broker-0] [zb-actors-0] [ZeebePartition-1] DEBUG io.camunda.zeebe.engine.state.migration - Copying entry with key[DbByte{bytes=���$MessageSubscriptionSentTimeMigration}] from column family [50] DEPRECATED_SIGNAL_SUBSCRIPTION_BY_NAME_AND_KEY to column family [56] MIGRATIONS_STATE
18:26:52.270 [Broker-0] [zb-actors-0] [ZeebePartition-1] DEBUG io.camunda.zeebe.engine.state.migration - Deleting entry with key[DbByte{bytes=���$MessageSubscriptionSentTimeMigration}] from column family [50] DEPRECATED_SIGNAL_SUBSCRIPTION_BY_NAME_AND_KEY
18:26:52.270 [Broker-0] [zb-actors-0] [ZeebePartition-1] DEBUG io.camunda.zeebe.engine.state.migration - Copying entry with key[DbByte{bytes=���+ProcessMessageSubscriptionSentTimeMigration}] from column family [50] DEPRECATED_SIGNAL_SUBSCRIPTION_BY_NAME_AND_KEY to column family [56] MIGRATIONS_STATE
18:26:52.270 [Broker-0] [zb-actors-0] [ZeebePartition-1] DEBUG io.camunda.zeebe.engine.state.migration - Deleting entry with key[DbByte{bytes=���+ProcessMessageSubscriptionSentTimeMigration}] from column family [50] DEPRECATED_SIGNAL_SUBSCRIPTION_BY_NAME_AND_KEY
18:26:52.271 [Broker-0] [zb-actors-0] [ZeebePartition-1] DEBUG io.camunda.zeebe.engine.state.migration - ColumnFamilyPrefixCorrectionMigration migration completed in 2 ms.
18:26:52.271 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Finished ColumnFamilyPrefixCorrectionMigration migration (16/18)
18:26:52.271 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Skipping MultiTenancySignalSubscriptionStateMigration migration (17/18).  It was determined it does not need to run right now.
18:26:52.271 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Starting JobBackoffRestoreMigration migration (18/18)
18:26:52.271 [Broker-0] [zb-actors-0] [ZeebePartition-1] DEBUG io.camunda.zeebe.engine.state.migration - JobBackoffRestoreMigration migration completed in 0 ms.
18:26:52.271 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Finished JobBackoffRestoreMigration migration (18/18)
18:26:52.271 [Broker-0] [zb-actors-0] [ZeebePartition-1] INFO  io.camunda.zeebe.engine.state.migration - Completed processing of migration tasks (use LogLevel.DEBUG for more details) ... 
18:26:52.271 [Broker-0] [zb-actors-0] [ZeebePartition-1] DEBUG io.camunda.zeebe.engine.state.migration - Executed 5 migration tasks: ProcessInstanceByProcessDefinitionMigration, JobTimeoutCleanupMigration, JobBackoffCleanupMigration, ColumnFamilyPrefixCorrectionMigration, JobBackoffRestoreMigration
```

</p>
</details> 

After taking a snapshot, we can also see this take effect in the runtime via zdb:

<details><summary>zdb output</summary>
<p>

State after startup, before a snapshot is taken. Note the 6 entries in the `DEPRECATED_SIGNAL_SUBSCRIPTION_BY_NAME_AND_KEY` column family.

```
zdb state -p raft-partition/partitions/1/runtime | jq
{
  "DEFAULT": 1,
  "KEY": 1,
  "EXPORTER": 2,
  "DEPRECATED_SIGNAL_SUBSCRIPTION_BY_NAME_AND_KEY": 6,
  "PROCESS_VERSION": 1,
  "PROCESS_CACHE": 1,
  "PROCESS_CACHE_BY_ID_AND_VERSION": 1,
  "PROCESS_CACHE_DIGEST_BY_ID": 1
}
```

Runtime state after snapshot is taken. Note the 6 entries in the `MIGRATIONS_STATE` column family.

```
❯ zdb state -p raft-partition/partitions/1/runtime | jq
{
  "DEFAULT": 1,
  "KEY": 1,
  "MIGRATIONS_STATE": 6,
  "PROCESS_VERSION": 1,
  "PROCESS_CACHE": 1,
  "PROCESS_CACHE_BY_ID_AND_VERSION": 1,
  "PROCESS_CACHE_DIGEST_BY_ID": 1
}
```

Exactly why a snapshot must be created before zdb can detect the changes in the runtime is not clear to me.

</p>
</details> 

## Related issues

closes #15804

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [x] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/main/java/io/camunda/zeebe/engine/state/migration/to_8_4/DbColumnFamilyCorrectionMigrationState.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  public void correctColumnFamilyPrefix() {
+    signalNameAndSubscriptionKeyColumnFamily.forEach(
+        (key, value) -> {
+          final DbString signalName = new DbString();
+          final DbLong subscriptionKey = new DbLong();
+          final DbCompositeKey<DbString, DbLong> signalNameAndSubscriptionKey =
+              new DbCompositeKey<>(signalName, subscriptionKey);
+          try {
+            signalNameAndSubscriptionKey.wrap(key.getDirectBuffer(), 0, key.getLength());
+          } catch (final Exception e) {
```
Comment (Zelldon) [inline engine/src/main/java/io/camunda/zeebe/engine/state/migration/to_8_4/DbColumnFamilyCorrectionMigrationState.java:62]: Just small input: I'm not 100% sure whether this is sufficient enough. But of course I also don't know what you have planned for the other types of CFs. It might be worth to check the length, and fields which have been written in the key. 

Why I'm unsure?

Because the composite keys are not writing the length itself and not what they types to be expected https://github.com/camunda/zeebe/blob/main/zb-db/src/main/java/io/camunda/zeebe/db/impl/DbCompositeKey.java#L56-L60

Let's say you have a [String](https://github.com/camunda/zeebe/blob/main/zb-db/src/main/java/io/camunda/zeebe/db/impl/DbString.java#L48-L54) and [Long](https://github.com/camunda/zeebe/blob/main/zb-db/src/main/java/io/camunda/zeebe/db/impl/DbLong.java#L37) as composite. 

The key would look like this: <CF, <STR_LEN, STRING, LONG>>

If I have a long long

<CF, <LONG, LONG>>

One problem we would have if the string len is zero then it is exact the same as the above. This means it would not fail to parse.

If I expect a <CF, <STR_LEN, String>>

I would still be able to parse the first composite key <CF, <STR_LEN, STRING, LONG>>, etc.

Not sure whether we have more examples. But I think you get the idea.

**Review by saig0 — APPROVED** [review]
@korthout looks good from my point of view. :+1: 

I love the new tests. :1st_place_medal: 

However, I don't have the full picture of the issue. I didn't check all the details. :sweat_smile:

**Comment by korthout** [comment]
Thanks @saig0 ❤️ 

>However, I don't have the full picture of the issue. I didn't check all the details. 

👍 That's alright. Thanks for being clear about it. IMO, we don't need a full review from a single person if we can get some partial reviews. Any comments about these changes are useful.

File: engine/src/main/java/io/camunda/zeebe/engine/state/migration/to_8_4/corrections/ColumnFamily48Corrector.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    messageStatsColumnFamily =
+        zeebeDb.createColumnFamily(
+            CF_POSSIBLE_TARGET,
+            transactionContext,
+            messagesDeadlineCountKey,
+            messagesDeadlineCount);
+  }
+
+  public void correctColumnFamilyPrefix() {
+    recoveryDmnDecisionKeyByDecisionIdAndVersion.forEach(
```
Comment (npepinpe) [inline engine/src/main/java/io/camunda/zeebe/engine/state/migration/to_8_4/corrections/ColumnFamily48Corrector.java:71]: Iterating over all key value pairs will lead to OOM if you have more state than memory, esp. since all the pairs end up allocated twice. I have a pending PR which solves this issue, but it's not merged yet.

See https://github.com/camunda/zeebe/pull/16090, specifically the poorly named class MemoryBoundedColumnIteration and how it's used. Obviously the drain method won't work here since you don't want to delete all existing pairs, but we can adapt it.

We could merge your PR first, then mine, then fix it in a follow up as long as we don't release in between

Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/state/migration/to_8_4/corrections/ColumnFamily48Corrector.java:71]: Do you think this is blocking a patch release for 8.4?

Comment (npepinpe) [reply engine/src/main/java/io/camunda/zeebe/engine/state/migration/to_8_4/corrections/ColumnFamily48Corrector.java:71]: It will cause customers who have to migrate a lot of data here to go out of memory. What do you think the likelihood of having lots of data here, or at least more than fits in memory, is? I guess it's a bit different since we're not moving _everything_ (as opposed to the other issue), but it could still be a lot?

Comment (npepinpe) [reply engine/src/main/java/io/camunda/zeebe/engine/state/migration/to_8_4/corrections/ColumnFamily48Corrector.java:71]: Do you want to sync? I need a reviewer on our PR (Berkay's and mine) anyway :upside_down_face:

Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/state/migration/to_8_4/corrections/ColumnFamily48Corrector.java:71]: The message stats contains a single entry, and the migration state contains a maximum of 20 entries.

But, there can be many entries in these CFs:
- one for each process instance
- one for each signal subscription
- one for each decision requirements
- one for each decision (there can be multiple per decision requirements)

So, let's fix this.

How does the MemoryBoundedColumnIteration ensure that it is continued once the memory is full?

Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/state/migration/to_8_4/corrections/ColumnFamily48Corrector.java:71]: After discussing this in more detail, we decided against adding the ability to chunk these migrations when going out of memory. This is because it's highly unlikely that this migration task will go out of memory.

The memory is increased per operation (i.e. PUT, DELETE) in the transaction. The increase is based on the number of bytes in these operations:
- only the key for DELETE
- the key + the value for PUT

The affected column families may contain many entries, but only a few must be moved/deleted. A move would be a PUT **and** a DELETE. The following entries would need to be moved/deleted:
- In 48, only one entry can be moved to `MESSAGE_STATS`
- In 49, one entry could be moved per active process instance to `PROCESS_INSTANCE_KEY_BY_DEFINITION_KEY`
- In 50, only a maximum of 20 entries can be moved to `MIGRATIONS_STATE`

So, only 49 has some actual number of entries to move. Each move is a DELETE of the key and a PUT of the key and value. The key combines the process definition key and the process instance key (twice one long of 8 bytes, so 16 bytes total). The value is a NilValue which is represented by a single byte.

So, a DELETE is just 16 bytes, and a PUT is just 17 bytes. Making each move a total of 33 bytes. At 1 million process instances, this would only take 33MB.

**Comment by korthout** [comment]
Thanks, everyone for your reviews and contributions! 🚀 

As we saw it's highly unlikely that this migration task reaches an OOM, it's ready to be merged.
